### PR TITLE
Recipient field and enhanced tests

### PR DESCRIPTION
**Pull request by 0age**
This PR contains an interface change to `fulfillAdvancedOrder` and `fulfillAvailableAdvancedOrders`: it adds a `recipient` parameter that the fulfiller can supply and that will receive all offered items on the fulfilled orders (with `address(0)` indicating to set the caller as the recipient, i.e. the default / original behavior). It also modifies the `OrderFulfilled` event, renaming the `fulfiller` parameter to `recipient` (and setting it to the recipient of the fulfilled offer items, or leaving it empty when orders are matched). The PR also contains a number of improvements to forge tests, including tests for fixes to various findings from recent code reviews.

### CODE REVIEW

File: contracts/lib/OrderFulfiller.sol [hunk 1]
```diff
@@ -112,16 +114,16 @@ contract OrderFulfiller is
             orderParameters,
             fillNumerator,
             fillDenominator,
-            orderParameters.conduitKey,
-            fulfillerConduitKey
+            fulfillerConduitKey,
+            recipient
```
[the remark this answers is not in the document]
Comment (gpersoon) [reply contracts/lib/OrderFulfiller.sol:118]: Actually `orderParameters.conduitKey` is removed from the parameter list.
And `orderParameters.conduitKey` seems to be retrieved from `orderParameters`, which is also passed.

**Comment by gpersoon** [comment]
Shouldn't the `msg.sender` in `_transferConsiderationItem` be replaced with `recipient`?
https://github.com/OpenSeaProtocol/seaport/blob/6cdc451d052b6dec7cfd819fb7d79be5588cbe2b/contracts/lib/OrderFulfiller.sol#L363
```solidity
              // Transfer item from caller to recipient specified by the item.
                _transferConsiderationItem(
                    considerationItem,
                    msg.sender,   // shouldn't this be recipient?
                    fulfillerConduitKey,
                    accumulator
                );
```

File: contracts/lib/Consideration.sol [hunk 2]
```diff
@@ -232,6 +238,7 @@ contract Consideration is ConsiderationInterface, OrderCombiner {
                 offerFulfillments,
                 considerationFulfillments,
                 fulfillerConduitKey,
+                msg.sender,
```
Comment (axic) [inline contracts/lib/Consideration.sol:241]: Why wasn't `recipient` added to the regular order fulfillers (`fulfillAvailableOrders` and `fulfillOrder`) only to the advanced ones?

**Comment by 0age** [comment]
> Shouldn't the `msg.sender` in `_transferConsiderationItem` be replaced with `recipient`? https://github.com/OpenSeaProtocol/seaport/blob/6cdc451d052b6dec7cfd819fb7d79be5588cbe2b/contracts/lib/OrderFulfiller.sol#L363
> 
> ```solidity
>               // Transfer item from caller to recipient specified by the item.
>                 _transferConsiderationItem(
>                     considerationItem,
>                     msg.sender,   // shouldn't this be recipient?
>                     fulfillerConduitKey,
>                     accumulator
>                 );
> ```

The caller still has to _pay_ for each spent item; otherwise a caller could name a recipient and spend their approved tokens (as there’s no access control on who the recipient is). So it’s really just a way to set the recipient for the offer items on the fulfilled order(s), not to declare the payer of the consideration items on the fulfilled order.

File: contracts/lib/OrderFulfiller.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (0age) [reply contracts/lib/OrderFulfiller.sol:118]: Yeah the offerer’s conduit key is read from parameters now to deal with stack pressure.

**Pull request merged**